### PR TITLE
feat: finish pagination

### PR DESCRIPTION
**Pull request by Doozers**
Signed-off-by: ismael FALL <ismael.fall@epitech.eu>

# Description

FIX #868 

## Contributors Checklist

- [x] Added new tests, or not needed, or not feasible
- [ ] Provided an example (e.g. screenshot) to aid review or the PR is self-explanatory
- [ ] Updated the official documentation or not needed
- [x] No breaking changes were made, or a `BREAKING CHANGE: xxx` message was included in the description
- [x] Added references to related issues and PRs
- [x] Provided any useful hints for running manual tests

## Maintainers Checklist

- [ ] Checked that the author followed the guidelines in `CONTRIBUTING.md`
- [x] Checked the conventional-commit (especially PR title and verb, presence of `BREAKING CHANGE:` in the body)
- [ ] Ensured that this PR is not a significant change or confirmed that the review/consideration process was appropriate for the change

### CODE REVIEW

File: examples/gno.land/p/demo/paginate/paginate.gno [hunk 1]
```diff
@@ -37,60 +138,65 @@ func RenderTreeByPage(
 	}
 
 	str := ""
-	numpages := (t.Size() / size) + 1
+	numPages := (t.Size() / size)
+	if numPages < 1 {
+		panic("invalid tree size to RenderTreeByPage")
```
Comment (moul) [inline examples/gno.land/p/demo/paginate/paginate.gno:143]:
```suggestion
		return ""
```

seems to be a more appropriate behavior, nope?

File: examples/gno.land/p/demo/paginate/paginate_test.gno [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+	mock.Set("2", &testItems{content: "content 2"})
+	mock.Set("3", &testItems{content: "content 3"})
+	mock.Set("4", &testItems{content: "content 4"})
+	mock.Set("5", &testItems{content: "content 5"})
+	mock.Set("6", &testItems{content: "content 6"})
+	mock.Set("7", &testItems{content: "content 7"})
+	mock.Set("8", &testItems{content: "content 8"})
+}
+
+func TestPaginate(t *testing.T) {
```
Comment (moul) [inline examples/gno.land/p/demo/paginate/paginate_test.gno:38]: can you make sure to also test all the cases like `pageFn == nil`, thx.

File: examples/gno.land/p/demo/paginate/paginate.gno [hunk 3]
```diff
@@ -12,22 +19,116 @@ type renderer interface {
...
+		"`":  "%60",
+		"{":  "%7B",
+		"|":  "%7C",
+		"}":  "%7D",
+		"~":  "%7E",
+		"\n": "%0A",
+	}
+	for k, v := range schars {
+		s = strings.Replace(s, k, v, -1)
+	}
```
Comment (moul) [inline examples/gno.land/p/demo/paginate/paginate.gno:54]:
```suggestion
	escapedURL := url.QueryEscape(s)
	escapedURL = strings.ReplaceAll(escapedURL, "%2F", "/") // Unescape forward slashes
	return escapedURL
```

not sure if I miss something?

File: examples/gno.land/p/demo/paginate/paginate.gno [hunk 4]
```diff
@@ -12,22 +19,116 @@ type renderer interface {
 // TODO this function belongs somewhere convenient.
```
Comment (moul) [inline examples/gno.land/p/demo/paginate/paginate.gno:19]: btw, a proper place would be a new `gno.land/p/demo/md` package with such generic helpers related to markdown.

feel free to move it right now.

File: examples/gno.land/p/demo/paginate/paginate.gno [hunk 5]
```diff
@@ -1,6 +1,13 @@
...
 
-import "strconv"
+import (
+	"strconv"
+	"strings"
+
+	"gno.land/p/demo/avl"
+)
+
+const PAGINATE_GAP = 3
```
Comment (moul) [inline examples/gno.land/p/demo/paginate/paginate.gno:10]:
```suggestion
const PaginateGap = 3
```

more conventional

File: examples/gno.land/p/demo/paginate/paginate.gno [hunk 6]
```diff
@@ -12,22 +19,116 @@ type renderer interface {
...
+	}
+	return s
 }
 
-var itoa = strconv.Itoa
+func navigation(numPages int, current int, reverse bool, pageFn pageFn) string {
+	str := ""
+
+	i := 1
+	for ; i <= PAGINATE_GAP && i < current; i++ {
```
Comment (moul) [inline examples/gno.land/p/demo/paginate/paginate.gno:62]:
```suggestion
	for i := 1 ; i <= PAGINATE_GAP && i < current; i++ {
```

Comment (moul) [reply examples/gno.land/p/demo/paginate/paginate.gno:62]:
```suggestion
	upperLimit := min(PaginateGap + 1, current)
	for i := 1; i < upperLimit; i++ {
```

easier to read

File: examples/gno.land/p/demo/paginate/paginate.gno [hunk 7]
```diff
@@ -12,22 +19,116 @@ type renderer interface {
...
+		pageurl := pageFn(i, numPages, reverse)
+		str += "[" + strconv.Itoa(i) + "](" +
+			escMDURL(pageurl) + ") "
```
Comment (moul) [inline examples/gno.land/p/demo/paginate/paginate.gno:65]:
```suggestion
		str += md.Link(strconv.Itoa(i),  pageurl)
```

if you go with a new `md` library with `md.EscapeURL(s string)` and `md.Link(title, url string)`

**Comment by moul** [comment]
Also, please fix the CI issue, which can be identified by running `gno test -precompile`.

File: examples/gno.land/p/demo/paginate/paginate.gno (hunk 3 above)
Comment (Doozers) [reply examples/gno.land/p/demo/paginate/paginate.gno:54]: I don't see any url package for Gno ?

Comment (moul) [reply examples/gno.land/p/demo/paginate/paginate.gno:54]: ok, so makes sense to keep the current implementation for now, and maybe keep my proposal for later

File: examples/gno.land/p/demo/paginate/paginate.gno (hunk 6 above)
Comment (Doozers) [reply examples/gno.land/p/demo/paginate/paginate.gno:62]: keep the index along the way over the range that's why I don't declare it in the scope of a loop.

For the `upperLimit` variable I tried to do it in a first way but didn't see any `math.Min` do you think it is valuable to do a PR to add it for the numeric values ?

Comment (Doozers) [reply examples/gno.land/p/demo/paginate/paginate.gno:62]: inside the Math package

Comment (moul) [reply examples/gno.land/p/demo/paginate/paginate.gno]: no, something like this is better

```go
func min(a, b int) int {...}
```

File: examples/gno.land/p/demo/md/md.gno [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+		"]":  "%5D",
+		"^":  "%5E",
+		"_":  "%5F",
+		"`":  "%60",
+		"{":  "%7B",
+		"|":  "%7C",
+		"}":  "%7D",
+		"~":  "%7E",
+		"\n": "%0A",
+	}
```
Comment (ajnavarro) [inline examples/gno.land/p/demo/md/md.gno:45]: I think this might be better being outside the function for performance reasons when executed on the Gno VM.

**Review by ajnavarro — COMMENTED** [review]
LGTM

File: examples/gno.land/p/demo/md/md.gno [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+	for k, v := range schars {
+		s = strings.Replace(s, k, v, -1)
+	}
+	return s
+}
+
+func Link(text, url string) string {
```
Comment (tbruyelle) [inline examples/gno.land/p/demo/md/md.gno:52]: You should probably use the existing `gno.land/p/demo/ui.Link` instead of creating a new one.

File: examples/gno.land/p/demo/paginate/paginate.gno [hunk 10]
```diff
@@ -37,60 +103,69 @@ func RenderTreeByPage(
...
+				str += "\n"
+				str += "\n"
+
+				return true
+			},
+		)
 	}
 
+	// render pagination again if there are pages.
+	str += navigation(numPages, page, reverse, pageFn)
```
Comment (tbruyelle) [inline examples/gno.land/p/demo/paginate/paginate.gno:167]: You should re-use the last call to `navigation()` to save some CPU cycles.

File: examples/gno.land/p/demo/paginate/paginate.gno [hunk 11]
```diff
@@ -37,60 +103,69 @@ func RenderTreeByPage(
...
+		t.IterateByOffset(offset, size,
+			func(key string, value interface{}) bool {
+				r := value.(renderer)
+				str += "\n"
+				str += r.Render()
+				str += "\n"
+				str += "\n"
+
+				return true
+			},
```
Comment (tbruyelle) [inline examples/gno.land/p/demo/paginate/paginate.gno:162]: repeated code (see lines 113-130)

File: examples/gno.land/p/demo/paginate/paginate.gno [hunk 12]
```diff
@@ -9,25 +16,84 @@ type renderer interface {
...
+// var itoa = strconv.Itoa
 
 // pageFn: path function to render paginattion links. if nil, no links are
 // rendered. result must start with ':'.
 // t: the tree with elements. each value must implement Render() string.
 // page: page number starting from 1.
 // size: page size.
-// reverse: reverse order. // XXX implement
+// reverse: reverse order.
 func RenderTreeByPage(
```
Comment (tbruyelle) [inline examples/gno.land/p/demo/paginate/paginate.gno:92]: What about exposing a `p/demo/ui.Pager` struct instead ? (cc @moul )

Comment (moul) [reply examples/gno.land/p/demo/paginate/paginate.gno:92]: if generic yes, else, we can make UI extendable with composition. 👍

File: examples/gno.land/p/demo/md/md.gno (hunk 9 above)
Comment (Doozers) [reply examples/gno.land/p/demo/md/md.gno:52]: :+1:

File: examples/gno.land/p/demo/paginate/paginate.gno (hunk 10 above)
Comment (Doozers) [reply examples/gno.land/p/demo/paginate/paginate.gno:167]: :+1:

File: examples/gno.land/p/demo/paginate/paginate.gno (hunk 12 above)
Comment (moul) [reply examples/gno.land/p/demo/paginate/paginate.gno:117]: Keep pagination separate from `ui`. Consider:

* `ui`: A minimal, independent UI for future composition.
* `paginate`: A standalone pagination engine.
* `gnoostrap`: An opinionated collection utilizing `ui`, `paginate` and other packages, akin to Twitter's Bootstrap.

File: examples/gno.land/p/demo/paginate/paginate.gno [hunk 13]
```diff
@@ -37,60 +103,69 @@ func RenderTreeByPage(
 	}
 
 	str := ""
-	numpages := (t.Size() / size) + 1
+	numPages := (t.Size() / size)
+	if numPages < 1 {
+		return ""
```
[the remark this answers is not in the document]
Comment (Doozers) [reply examples/gno.land/p/demo/paginate/paginate.gno:135]: I mean if the size of a single page is greater than the size of the whole tree it is not possible then to render any pages ? @tbruyelle

[the remark this answers is not in the document]
Comment (tbruyelle) [reply examples/gno.land/p/demo/paginate/paginate.gno]: Let's say the tree contains 5 items and the page size is 10, then `(t.Size()/size)` will give `0`, and so your code will panic. Why not rendering the 5 items of the tree instead ?

[the remark this answers is not in the document]
Comment (Doozers) [reply examples/gno.land/p/demo/paginate/paginate.gno]: Ok right

File: examples/gno.land/p/demo/paginate/paginate_test.gno [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+	const testString = `[1](%3A//here%20is%20the%20url/1) *2* [3](%3A//here%20is%20the%20url/3) ... [6](%3A//here%20is%20the%20url/6) [7](%3A//here%20is%20the%20url/7) [8](%3A//here%20is%20the%20url/8) 
+
+CONTENT OF TESTITEM: { content 2 }
+
+[1](%3A//here%20is%20the%20url/1) *2* [3](%3A//here%20is%20the%20url/3) ... [6](%3A//here%20is%20the%20url/6) [7](%3A//here%20is%20the%20url/7) [8](%3A//here%20is%20the%20url/8) 
+`
+	res := RenderTreeByPage(mock, paginTest, 2, 1, true)
```
Comment (tbruyelle) [inline examples/gno.land/p/demo/paginate/paginate_test.gno:45]: Can you add 2 more cases in the tests, related to the size parameter:
- size > 1 and <= `mock.Size()`
- size > `mock.Size()`

**Comment by thehowl** [comment]
@Doozers just a friendly ping asking if you intend to resolve conversations and finish the PR. Once you do that, I'd also recommend re-creating this PR to merge into `master` rather than `dev/jae/paginate` -- I'm closing the other PR as you took the torch to finish it here :)

**Comment by Doozers** [comment]
sure sorry didn't see new comments 👍

File: examples/gno.land/p/demo/paginate/paginate_test.gno (hunk 14 above)
Comment (Doozers) [reply examples/gno.land/p/demo/paginate/paginate_test.gno:45]: done for the first test but I don't see any way to test a panic so the second is impossible to test
